test/FunctionsV2/AppAuthenticationCredentialsFactoryTests.cs

```diff
@@ -0,0 +1,174 @@
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using Microsoft.WindowsAzure.Storage.Auth;
+using Moq;
+using Xunit;
+using Xunit.Sdk;
+
+namespace WebJobs.Extensions.DurableTask.Tests.V2
+{
+    public class AppAuthenticationCredentialsFactoryTests
+    {
+        [Theory]
+        [Trait("Category", PlatformSpecificHelpers.TestCategory)]
+        [InlineData(null, null, null, null, null, default(StoreLocation), null)]
+        [InlineData("managedidentity", null, null, null, null, default(StoreLocation), "RunAs=App")]
+        [InlineData("managedidentity", "UserA", null, null, null, default(StoreLocation), "RunAs=App;AppId=UserA")]
+        [InlineData(null, null, "MyTenant", "OpenSesame", null, default(StoreLocation), null)]
+        [InlineData(null, "UserB", null, "OpenSesame", null, default(StoreLocation), null)]
+        [InlineData(null, "UserB", "MyTenant", null, null, default(StoreLocation), null)]
+        [InlineData(null, "UserB", "MyTenant", "OpenSesame", null, default(StoreLocation), "RunAs=App;AppId=UserB;TenantId=MyTenant;AppKey=OpenSesame")]
+        [InlineData(null, null, "MyOtherTenant", null, "MyThumbprint", StoreLocation.LocalMachine, null)]
+        [InlineData(null, "UserC", null, null, "MyThumbprint", StoreLocation.LocalMachine, null)]
+        [InlineData(null, "UserC", "MyOtherTenant", null, null, StoreLocation.LocalMachine, null)]
+        [InlineData(null, "UserC", "MyOtherTenant", null, "MyThumbprint", StoreLocation.LocalMachine, "RunAs=App;AppId=UserC;TenantId=MyOtherTenant;CertificateThumbprint=MyThumbprint;CertificateStoreLocation=LocalMachine")]
+        [InlineData(null, "UserD", "MyThirdTenant", null, "MyOtherThumbprint", StoreLocation.CurrentUser, "RunAs=App;AppId=UserD;TenantId=MyThirdTenant;CertificateThumbprint=MyOtherThumbprint;CertificateStoreLocation=CurrentUser")]
+        public void GetConnectionString(string credential, string appId, string tenantId, string clientSecret, string certificate, StoreLocation certificateStoreLocation, string expected)
+        {
+            var options = new AzureIdentityOptions
+            {
+                Certificate = certificate,
+                ClientId = appId,
+                ClientCertificateStoreLocation = certificateStoreLocation,
+                ClientSecret = clientSecret,
+                Credential = credential,
+                TenantId = tenantId,
+            };
+
+            Assert.Equal(expected, AppAuthenticationCredentialsFactory.GetConnectionString(options));
+        }
+
+        [Fact]
+        [Trait("Category", PlatformSpecificHelpers.TestCategory)]
+        public async Task NoConnection()
+        {
+            using var factory = new AppAuthenticationCredentialsFactory(NullLoggerFactory.Instance);
+            Assert.Null(await factory.CreateAsync(new AzureIdentityOptions()));
+        }
+
+        [Fact]
+        [Trait("Category", PlatformSpecificHelpers.TestCategory)]
+        public async Task CreateAsync()
```

Comment:
Is it possible to create some tests with renewingCalls and renewedCalls > 1? The renew interval can be super small. I think it's just a timer that runs so it should be pretty deterministic since you're mocking calls. We can also test the dispose logic to make sure it is not renewing after we should have disposed.

Reply:
Yes, I've added these scenarios